### DISCUSSION TITLE
BMC- Relay Rotation - Skipping Case

### DISCUSSION THREAD

**Discussion by quangtran8588** [icon-project/btp#27] [Q&A] [answered]
Hi jspark-icon,
- As previously mentioned in another discussion: 
"When becoming turn, BMR could delay to relay 'BlockUpdate' until end of term. but BMR should relay 'delayed - BlockUpdates' and 'ReceiptProofs' immediately at BTP-Message creation for reduce latency. in this case, if BMR doesn't relay on time ('delay_limit'), then rotate to next."

When *rotate_term* is 15 and *delay_limit* is 3, it means that BMR is allowed to delay to relay 'BlockUpdate' within [x, x+15], but it requires to relay 'Receipt_Proof' within [x, x+3]. Am I correct? If my undestanding is correct, what would be an expected handling behavior when one BMR is failed to relay 'Receipt_Proof' within [x, x+3]?
Case 1: from [x+3, x+15], another Relay will take place to relay BTP Message
Case 2: from [x+3, x+15], there will be no Relay taking responsibility to relay BTP Message. Then, a next rotation round, a next Relay will cover a skipping BTP Message by relaying (skipping 'BlockUpdate' and 'ReceiptProof', new 'BlockUpdate' and 'ReceiptProof')
Please correct me if I misunderstood

- Please explain logic of using *skip_count* when BTP message includes 'ReceiptProof'
```
def rotate(self, current_height: int, relay_msg_height: int, has_msg: bool = False) -> BTPRelay:
        rotate_term = self.rotate_term()
        if rotate_term > 0:
            if has_msg:
                guess_height = self.rx_height + math.ceil((relay_msg_height - self.rx_height_src) / self._scale()) - 1
                if guess_height > current_height:
                    guess_height = current_height
                rotate_count = math.ceil((guess_height - self.rotate_height)/rotate_term)
                if rotate_count < 0:
                    rotate_count = 0
                base_height = self.rotate_height + ((rotate_count-1) * rotate_term)
                skip_count = math.ceil((current_height - guess_height)/self.delay_limit) - 1
                if skip_count > 0:
                    rotate_count += skip_count
                    base_height = current_height
                self.rx_height = current_height
                self.rx_height_src = relay_msg_height
                return self._rotate(rotate_count, base_height)
            else:
                rotate_count = math.ceil((current_height - self.rotate_height)/rotate_term)
                base_height = self.rotate_height + ((rotate_count-1) * rotate_term)
                return self._rotate(rotate_count, base_height)
        else:
            return None
```

I'm really appreciate if you can help me to understand an implementation. Thanks for your time.

**Comment by cow-hs** [comment, marked as answer]
I will attach a description of the Round-Robin schedule, Fail-over situation for BMC Rotate.
Please note.

[https://drive.google.com/file/d/1prIOmTudoWLV0BzqzR0aV-aJ9quJyhhQ/view?usp=sharing] (round-robin schedule)
[https://drive.google.com/file/d/1f5z3I32VtUxRpHVrqllEq1EH_fuF2N-3/view?usp=sharing] (fail-over)

**Reply by quangtran8588** [reply]
Many thanks. It's clear for me now
